src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/KeyValuePairModelBinder.cs

```diff
@@ -17,9 +17,13 @@ public sealed class KeyValuePairModelBinder<TKey, TValue> : IModelBinder
                 throw new ArgumentNullException(nameof(bindingContext));
             }
 
-            ModelBindingHelper.ValidateBindingContext(bindingContext,
-                                                      typeof(KeyValuePair<TKey, TValue>),
-                                                      allowNullModel: true);
+            if (bindingContext.ModelType != typeof(KeyValuePair<TKey, TValue>))
+            {
+                var message = Resources.FormatModelBinderUtil_ModelTypeIsWrong(
+                    bindingContext.ModelType,
+                    typeof(KeyValuePair<TKey, TValue>));
+                throw new ArgumentException(message, nameof(bindingContext));
+            }
```

Comment:
Will go with just returning. Also filed #4181 for other oddities related to behaving like these types are `public`.